File: src/main/java/rag/model/Chunk.java
package main.java.rag.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Chunk {
    private String docId;
    private String chunkId;
    private String text;
    private String sectionId;
    private int startOffset;
    private int endOffset;
    
    public Chunk() {
    }
    
    public Chunk(String docId, String chunkId, String text, String sectionId, int startOffset, int endOffset) {
        this.docId = docId;
        this.chunkId = chunkId;
        this.text = text;
        this.sectionId = sectionId;
        this.startOffset = startOffset;
        this.endOffset = endOffset;
    }
    
    public String getDocId() {
        return docId;
    }
    
    public void setDocId(String docId) {
        this.docId = docId;
    }
    
    public String getChunkId() {
        return chunkId;
    }
    
    public void setChunkId(String chunkId) {
        this.chunkId = chunkId;
    }
    
    public String getText() {
        return text;
    }
    
    public void setText(String text) {
        this.text = text;
    }
    
    public String getSectionId() {
        return sectionId;
    }
    
    public void setSectionId(String sectionId) {
        this.sectionId = sectionId;
    }
    
    public int getStartOffset() {
        return startOffset;
    }
    
    public void setStartOffset(int startOffset) {
        this.startOffset = startOffset;
    }
    
    public int getEndOffset() {
        return endOffset;
    }
    
    public void setEndOffset(int endOffset) {
        this.endOffset = endOffset;
    }
    
    public List<String> getSentences() {
        List<String> sentences = new ArrayList<>();
        if (text == null || text.isEmpty()) {
            return sentences;
        }
        
        String[] parts = text.split("(?<=[.!?])\\s+");
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty()) {
                sentences.add(trimmed);
            }
        }
        
        if (sentences.isEmpty() && !text.trim().isEmpty()) {
            sentences.add(text.trim());
        }
        
        return sentences;
    }
    
    public boolean contains(String term) {
        if (text == null || term == null) {
            return false;
        }
        return text.toLowerCase().contains(term.toLowerCase());
    }
    
    public String formatCitation() {
        return docId + ":" + sectionId + ":" + startOffset + "-" + endOffset;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Chunk chunk = (Chunk) o;
        return startOffset == chunk.startOffset &&
               endOffset == chunk.endOffset &&
               Objects.equals(docId, chunk.docId) &&
               Objects.equals(chunkId, chunk.chunkId) &&
               Objects.equals(text, chunk.text) &&
               Objects.equals(sectionId, chunk.sectionId);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(docId, chunkId, text, sectionId, startOffset, endOffset);
    }
    
    @Override
    public String toString() {
        return "Chunk{" +
               "docId='" + docId + '\'' +
               ", chunkId='" + chunkId + '\'' +
               ", sectionId='" + sectionId + '\'' +
               ", startOffset=" + startOffset +
               ", endOffset=" + endOffset +
               ", textLength=" + (text != null ? text.length() : 0) +
               '}';
    }
}

File: iteration1_py/src/model/chunk.py
from typing import List
import re

class Chunk:
   
    def __init__(self, doc_id: str = None, chunk_id: str = None, text: str = None, 
                 section_id: str = None, start_offset: int = 0, end_offset: int = 0):
       
        self.__doc_id = doc_id
        self.__chunk_id = chunk_id
        self.__text = text
        self.__section_id = section_id
        self.__start_offset = start_offset if start_offset is not None else 0
        self.__end_offset = end_offset if end_offset is not None else 0
    
    def get_doc_id(self) -> str:
        """Get the document ID."""
        return self.__doc_id
    
    def get_chunk_id(self) -> str:
        """Get the chunk ID."""
        return self.__chunk_id
    
    def get_text(self) -> str:
        """Get the chunk text."""
        return self.__text
    
    def get_section_id(self) -> str:
        """Get the section ID."""
        return self.__section_id
    
    def get_start_offset(self) -> int:
        """Get the start offset."""
        return self.__start_offset
    
    def get_end_offset(self) -> int:
        """Get the end offset."""
        return self.__end_offset
    
    def get_sentences(self) -> List[str]:
        sentences = []
        if not self.__text:
            return sentences

        parts = re.split(r'(?<=[.!?])\s+', self.__text)
        for part in parts:
            trimmed = part.strip()
            if trimmed:
                sentences.append(trimmed)

        if not sentences and self.__text.strip():
            sentences.append(self.__text.strip())

        return sentences
    
    def contains(self, term: str) -> bool:
        if not self.__text or not term:
            return False
        return term.lower() in self.__text.lower()
    
    def format_citation(self) -> str:
        return f"{self.__doc_id}:{self.__section_id}:{self.__start_offset}-{self.__end_offset}"
    
    def __str__(self):
        text_length = len(self.__text) if self.__text else 0
        return (
            f"Chunk{{docId='{self.__doc_id}', chunkId='{self.__chunk_id}', "
            f"sectionId='{self.__section_id}', startOffset={self.__start_offset}, "
            f"endOffset={self.__end_offset}, textLength={text_length}}}"
        )
